File: recursion/javascript/solutions/isSorted.js
/**
 * Solution for isSorted.
 */

const { EMPTY_LIST, Node, isEmpty, unprepend } = require('./linkedList');

function isSorted(list) {
  if (isEmpty(list)) {
    return true;
  }

  if (isEmpty(list.next)) {
    return true;
  }

  let [first, rest] = unprepend(list);
  let second = rest.value;

  return first <= second && isSorted(rest);
}

if (require.main === module) {
  console.assert(isSorted(Node.fromArray([1, 2, 3])) === true);
  console.assert(isSorted(Node.fromArray([1, 3, 2])) === false);
  console.assert(isSorted(EMPTY_LIST) === true);
  console.assert(isSorted(Node.fromArray([42])) === true);
  console.assert(isSorted(Node.fromArray([1, 1, 2, 3])) === true);

  console.log('All tests passed!');
}

module.exports = { isSorted };

File: recursion/python/solutions/is_sorted.py
"""Solution for is_sorted."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend


def is_sorted(lst):
    """
    Return True if lst is sorted in ascending order.

    is_sorted(1 -> 2 -> 3) => True
    is_sorted(1 -> 3 -> 2) => False
    is_sorted(()) => True
    is_sorted(42) => True (single element)
    """
    if is_empty(lst):
        return True

    if is_empty(lst.next):
        return True

    first, rest = unprepend(lst)
    second = rest.value

    return first <= second and is_sorted(rest)


if __name__ == '__main__':
    # Test cases
    assert is_sorted(Node.from_list([1, 2, 3])) == True
    assert is_sorted(Node.from_list([1, 3, 2])) == False
    assert is_sorted(EMPTY_LIST) == True
    assert is_sorted(Node.from_list([42])) == True
    assert is_sorted(Node.from_list([1, 1, 2, 3])) == True
    assert is_sorted(Node.from_list([3, 2, 1])) == False

    print("All tests passed!")
